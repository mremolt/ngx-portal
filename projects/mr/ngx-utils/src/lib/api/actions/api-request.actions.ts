import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Action } from '@ngrx/store';

import { ApiActions, ApiRequestConfig } from '../types';

export enum ApiRequestActionTypes {
  ApiRequest = '[API] Request',
  ApiError = '[API] Error',
}

export class ApiRequest<T = any> implements Action {
  public readonly type = ApiRequestActionTypes.ApiRequest;

  constructor(public payload: ApiRequestConfig<T>) {}
}

export class ApiError implements Action {
  public readonly type = ApiRequestActionTypes.ApiError;

  constructor(public payload: HttpErrorResponse) {}
}

export class ApiGetAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('GET', url);
    super({ request, actions });
  }
}

export class ApiPostAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('POST', url, body);
    super({ request, actions });
  }
}

export class ApiPutAction<T> extends ApiRequest {
  constructor(url: string, body: T | null, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('PUT', url, body);
    super({ request, actions });
  }
}

export class ApiDeleteAction<T> extends ApiRequest {
  constructor(url: string, actions: ApiActions<T>) {
    const request = new HttpRequest<T>('DELETE', url);
    super({ request, actions });
  }
}
