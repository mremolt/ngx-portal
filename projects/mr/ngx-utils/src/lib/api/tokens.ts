import { InjectionToken } from '@angular/core';

import { ApiTokenSelector } from './types';

export const API_TOKEN_SELECTOR = new InjectionToken<ApiTokenSelector>('TOKEN_SELECTOR');
