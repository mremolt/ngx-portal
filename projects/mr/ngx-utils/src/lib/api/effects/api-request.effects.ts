import { HttpClient, HttpRequest, HttpResponse } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Store, select } from '@ngrx/store';
import { EMPTY, concat, of } from 'rxjs';
import { catchError, filter, flatMap, map, share, take } from 'rxjs/operators';

import { APP_ENVIRONMENT } from '../../tokens';
import { Environment } from '../../types';
import { ApiError, ApiRequest, ApiRequestActionTypes } from '../actions/api-request.actions';
import { API_TOKEN_SELECTOR } from '../tokens';
import { ApiTokenSelector } from '../types';

export function buildApiHttpRequest<T>(
  baseRequest: HttpRequest<T>,
  env: Environment,
  token: string
) {
  return baseRequest.clone({
    url: `${env.apiUrl}/${baseRequest.url}`,
    setHeaders: { Authorization: `Bearer ${token}` },
  });
}

@Injectable()
export class ApiRequestEffects {
  @Effect()
  processApiRequest$ = this.actions$.pipe(
    ofType<ApiRequest>(ApiRequestActionTypes.ApiRequest),
    flatMap(action => {
      const baseRequest = action.payload.request;
      const actions = action.payload.actions;

      return this.store.pipe(
        select(this.tokenSelector.select),
        take(1),
        flatMap(token => {
          const req = buildApiHttpRequest(baseRequest, this.env, token);

          return concat(
            of(new actions.start()),
            this.http.request(req).pipe(
              filter<any>(res => res instanceof HttpResponse),
              map((res: HttpResponse<any>) => res.body),
              map(data => new actions.success(data)),
              catchError(e => of(new actions.error(e), new ApiError(e)))
            ),
            actions.complete ? of(new actions.complete()) : EMPTY
          );
        })
      );
    }),
    share()
  );

  constructor(
    private actions$: Actions,
    private http: HttpClient,
    private store: Store<any>,
    @Inject(API_TOKEN_SELECTOR) private tokenSelector: ApiTokenSelector,
    @Inject(APP_ENVIRONMENT) private env: Environment
  ) {}
}
