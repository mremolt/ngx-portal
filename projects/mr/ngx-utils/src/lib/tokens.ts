import { InjectionToken } from '@angular/core';

import { Environment } from './types';

export const APP_ENVIRONMENT = new InjectionToken<Environment>('APP_ENVIRONMENT');
