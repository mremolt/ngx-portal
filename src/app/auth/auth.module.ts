import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { AuthEffects } from './reducers/auth.effects';
import { AuthGuard } from './reducers/auth.guard';
import * as fromAuth from './reducers/auth.reducer';

@NgModule({
  imports: [
    CommonModule,
    StoreModule.forFeature('auth', fromAuth.reducer),
    EffectsModule.forFeature([AuthEffects]),
  ],
  declarations: [],
  providers: [AuthGuard],
})
export class AuthModule {}
