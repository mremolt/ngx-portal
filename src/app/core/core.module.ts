import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { InjectionToken, NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreRouterConnectingModule } from '@ngrx/router-store';
import { ActionReducerMap, META_REDUCERS, MetaReducer, StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';

import { environment } from '../../environments/environment';
import { AppEffects } from '../app.effects';
import { AuthModule } from '../auth/auth.module';
import { RootState, metaReducers, reducers } from '../reducers';
import { APP_ENVIRONMENT } from '../tokens';

export const REDUCER_TOKEN = new InjectionToken<ActionReducerMap<RootState>>('Registered Reducers');

export function getReducers() {
  return reducers;
}

export function getMetaReducers(): MetaReducer<RootState>[] {
  return metaReducers;
}

@NgModule({
  imports: [
    // Angular
    CommonModule,
    HttpClientModule,
    // ngrx
    StoreModule.forRoot(REDUCER_TOKEN),
    !environment.production ? StoreDevtoolsModule.instrument() : [],
    EffectsModule.forRoot([AppEffects]),
    StoreRouterConnectingModule.forRoot(),
    // Features
    AuthModule,
  ],
  declarations: [],
  providers: [
    {
      provide: REDUCER_TOKEN,
      useFactory: getReducers,
    },
    {
      provide: META_REDUCERS,
      useFactory: getMetaReducers,
    },
    {
      provide: APP_ENVIRONMENT,
      useValue: environment,
    },
  ],
})
export class CoreModule {}
