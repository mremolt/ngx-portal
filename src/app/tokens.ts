import { InjectionToken } from '@angular/core';

import { Environment } from '../environments/environment';

export const APP_ENVIRONMENT = new InjectionToken<Environment>('APP_ENVIRONMENT');
