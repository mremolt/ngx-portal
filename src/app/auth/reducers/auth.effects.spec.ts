import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { TestBed } from '@angular/core/testing';
import { provideMockActions } from '@ngrx/effects/testing';
import { ReplaySubject, Subject } from 'rxjs';

import { APP_ENVIRONMENT } from '../../tokens';
import { Authenticate, AuthenticateError, AuthenticateSuccess } from './auth.actions';
import { AuthEffects } from './auth.effects';

describe('AuthEffects', () => {
  let effects: AuthEffects;
  let actions$: Subject<any>;
  let httpMock: HttpTestingController;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [HttpClientTestingModule],
      providers: [
        AuthEffects,
        { provide: APP_ENVIRONMENT, useValue: { apiUrl: 'http://localhost:3001' } },
        provideMockActions(() => actions$),
      ],
    });

    httpMock = TestBed.get(HttpTestingController);
    effects = TestBed.get(AuthEffects);
  });

  it('should be created', () => {
    expect(effects).toBeTruthy();
  });

  describe('authenticate$', () => {
    beforeEach(() => {
      actions$ = new ReplaySubject(1);
      actions$.next(new Authenticate('test1@example.com', 'secret'));
    });

    it('emits AuthenticateSuccess on successful HTTP call', done => {
      effects.authenticate$.subscribe(successAction => {
        expect(successAction).toBeInstanceOf(AuthenticateSuccess);
        expect(successAction).toMatchSnapshot();
        done();
      });

      const req = httpMock.expectOne('http://localhost:3001/auth/login');
      expect(req.request.method).toEqual('POST');

      req.flush({ authToken: '1234567890' });
      httpMock.verify();
    });

    it('emits AuthenticateError on error HTTP call', done => {
      effects.authenticate$.subscribe(errorAction => {
        expect(errorAction).toBeInstanceOf(AuthenticateError);
        expect(errorAction).toMatchSnapshot();
        done();
      });

      const req = httpMock.expectOne('http://localhost:3001/auth/login');
      expect(req.request.method).toEqual('POST');

      req.flush('', { status: 403, statusText: 'Nix!' });
      httpMock.verify();
    });
  });
});
