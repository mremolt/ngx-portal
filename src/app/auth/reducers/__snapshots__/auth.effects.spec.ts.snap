// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthEffects authenticate$ emits AuthenticateError on error HTTP call 1`] = `
AuthenticateError {
  "payload": HttpErrorResponse {
    "error": null,
    "headers": HttpHeaders {
      "headers": Map {},
      "lazyUpdate": null,
      "normalizedNames": Map {},
    },
    "message": "Http failure response for http://localhost:3001/auth/login: 403 Nix!",
    "name": "HttpErrorResponse",
    "ok": false,
    "status": 403,
    "statusText": "Nix!",
    "url": "http://localhost:3001/auth/login",
  },
  "type": "[Auth] Authenticate Error",
}
`;

exports[`AuthEffects authenticate$ emits AuthenticateSuccess on successful HTTP call 1`] = `
AuthenticateSuccess {
  "payload": Object {
    "authToken": "1234567890",
  },
  "type": "[Auth] Authenticate Success",
}
`;
